### ISSUE TITLE
Crashes when malformed HTTP request are sent.

### ISSUE THREAD

**Issue by siktec-lab** [nomasystems/erf#77]
Hi, First of all - I like the idea and the ease of use.

I noticed in the rebar3 shell that whenever I send a malformed request, the system crashes, and the response is `Internal server error`.

### How to reproduce: (using the user's example)
```
GET http://localhost:8080/users/234sadf2342345sdf/ HTTP/1.1
content-type: application/json
X-API-KEY: api-key
```
**Notice the content type is `application/json` with no body, and it should not be present in a GET request**

**This will cause:**
```
=ERROR REPORT==== 6-Sep-2024::03:00:31.501000 ===
erf_router:handle_event/3 crashed error:function_clause
[{njson_decoder,decode,
                [undefined],
                [{file,"d:/projects/silentedge/api/erl/agentserver/_build/default/lib/njson/src/njson_decoder.erl"},
                 {line,49}]},
 {erf_router,preprocess,1,
             [{file,"d:/projects/silentedge/api/erl/agentserver/_build/default/lib/erf/src/erf_router.erl"},
              {line,800}]},
 {erf_router,handle_event,3,
             [{file,"d:/projects/silentedge/api/erl/agentserver/_build/default/lib/erf/src/erf_router.erl"},
              {line,160}]},
 {elli_http,handle_event,4,
            [{file,"d:/projects/silentedge/api/erl/agentserver/_build/default/lib/elli/src/elli_http.erl"},
             {line,814}]},
 {elli_http,execute_callback,1,
            [{file,"d:/projects/silentedge/api/erl/agentserver/_build/default/lib/elli/src/elli_http.erl"},
             {line,332}]},
 {elli_http,handle_request,4,
            [{file,"d:/projects/silentedge/api/erl/agentserver/_build/default/lib/elli/src/elli_http.erl"},
             {line,116}]},
 {elli_http,keepalive_loop,5,
            [{file,"d:/projects/silentedge/api/erl/agentserver/_build/default/lib/elli/src/elli_http.erl"},
             {line,83}]},
 {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,329}]}]
```

**And the response:**
```
HTTP/1.1 500 Internal Server Error
connection: close
content-length: 21

Internal server error
```

## I wonder:
 - Shouldn't we ignore the BODY for GET requests?
 - I saw in the source code that the json parsing has an error match - why is it still crashing?

**Comment by siktec-lab** [comment]
Thank you @javiergarea. I really appreciate the quick response and the fix. 

**TL;DR: I agree with your points, and it makes sense not to enforce the policy for everyone.**

HTTP 1.1, not forbidding a body in GET requests, is a good point; from my experience, it can lead to confusion and errors.
Many services do not support bodies in GET requests, which can cause errors when integrating with them.
I have encountered this issue with caching and some proxies, clients, and opinionated services.

That said, I agree with you that it should not be enforced for everyone but should be configurable.

I also agree that an empty body with the `content-type: application/json` should be treated as invalid. Funny enough, some may argue that an empty body is a valid JSON and should be treated as such (e.g. bullish datatype). But it is better to treat it as invalid.

Final thoughts:
1. It is a good idea to make it configurable. It will be an excellent addition to the project and can save some developers headaches when integrating/supporting some services.
2. I suggest abstracting the RawBody parsing to a separate module so that users can easily replace or extend it.
Currently, `<<application/json>>` is hardcoded in the `erf_router.erl` and automatically parses the body as JSON. It would be nice to have a way to replace this with a custom parser or disable it entirely and add more parsers as needed (e.g. XML, YAML, etc.)

I appreciate your work and the project. I like the approach and simplicity of it. Keep up the excellent work!

I am currently working on a project. I will definitely use this and contribute to it in the near future.

**Comment by javiergarea** [comment]
Hi @siktec-lab, thank you for your kind words!

> I suggest abstracting the RawBody parsing to a separate module so that users can easily replace or extend it.
Currently, <<application/json>> is hardcoded in the erf_router.erl and automatically parses the body as JSON. It would be nice to have a way to replace this with a custom parser or disable it entirely and add more parsers as needed (e.g. XML, YAML, etc.)

You’re right—this is something we could spend some time on. The `application/json` by default (note that is not enforced) was based on the assumption that most users would rely on JSON. However, providing an easy method to extend the preprocessing step with custom parsers is a great idea. While you can create a preprocess_middleware for this, adding a configurable option would certainly improve both ergonomics and semantics. I created a issue to track this enhancement (#79).

Closing this issue, as the crashes have been resolved.
